File: day2.py
import numpy as np
constrains = {
    "green" : 13,
    "blue" : 14,
    "red" : 12
}
acc = 0
acc2 = 0
def is_valid(game):
    s = 0
    num_to_color = game.split(',')
    for x in num_to_color:
        x = x[1:]
        num = int(x.split(' ')[0])
        color = x.split(' ')[1]
        s+=num
        if constrains[color] < num:
            return False
    return sum(list(constrains.values())) > s

def get_min_valid(games):
    min_valid = {"green" : 0,
                 "blue" : 0,
                 "red" : 0}
    for game in games:
        num_to_color = game.split(',')
        for x in num_to_color:
            x = x[1:]
            num = int(x.split(' ')[0])
            color = x.split(' ')[1]
            min_valid[color] = max(min_valid[color], num)
    return min_valid 
    
with open("day2.txt", 'r') as f:
    for line in f.readlines():
        id = int(line.split(':')[0].split(' ')[1])
        games = line.replace('\n','').split(':')[1].split(';')
        valid = True
        for game in games:
            valid &= is_valid(game)
        if valid:
            acc+= id
        acc2 += np.prod(list(get_min_valid(games).values()))
    print(acc)
    print(acc2)

File: day2.hs
import Data.List
import Debug.Trace (traceShow)
import Data.Char (isDigit)
constrains = [("blue", 14),("green", 13), ("red", 12)]
maxb = tsum constrains
tsum :: [(String, Int)] -> Int
tsum [] = 0
tsum ((_, i):xs) = i + tsum xs

isPossible :: [(String, Int)] -> [(String, Int)] -> Bool
isPossible [] [] = True
isPossible (x:xs) [] = isPossible xs []
isPossible [] (x:xs) = isPossible [] xs
isPossible ((color, outcome) : xs) ((color2, constrain) : xs2)
    | sum1 > maxb = False
    | color == color2 = (outcome <= constrain) && isPossible xs xs2
    | otherwise = isPossible ((color, outcome) : xs) xs2 && isPossible xs ((color2, constrain) : xs2)
    where
        sum1 = tsum  ((color, outcome) : xs)

parseGame :: [String] -> [(String, Int)]
parseGame  (w:(w2:xs))
    | null xs = [(w2, read w)]
    | otherwise = (w2, read w) : parseGame xs

sumPossible :: [String] -> Int
sumPossible games
    | null games = 0
    | isPossibleGame (filter (/=':') right) = read game_num + sumPossible xs
    | otherwise = sumPossible xs
    where
        (game:xs) = games
        (left, b:right) = break (==':') game
        game_num = filter isDigit left

isPossibleGame :: String -> Bool
isPossibleGame game
    | null game = True
    | null right = isPossible (parseGame (words left)) constrains
    | otherwise = isPossible (parseGame (words left)) constrains && isPossibleGame (drop 1 right)
    where
        (left, right) = break (==';') game

wordsWhen :: (Char -> Bool) -> String -> [String]
wordsWhen p s =  case dropWhile p s of
                      "" -> []
                      s' -> w : wordsWhen p s''
                            where (w, s'') = break p s'

parseGames :: [String] -> [[(String, Int)]]
parseGames (x:xs)
    | null xs = [concatMap (parseGame . words) individualGames]
    | otherwise = concatMap (parseGame . words) individualGames : parseGames xs
    where
        (left, b:right) = break (==':') x
        individualGames = wordsWhen (==';') right

getMinRequired :: [(String,Int)] -> (Int,Int,Int)
getMinRequired game = (minRed, minBlue, minGreen)
    where
        minRed = maximum (map snd (filter (\x -> fst x =="red") game))
        minBlue = maximum (map snd (filter (\x -> fst x =="blue") game))
        minGreen = maximum (map snd (filter (\x -> fst x =="green") game))

main :: IO ()
main = do
    contents <- readFile "day2.txt"
    let games = lines (filter (/=',') contents)
    print (sumPossible games)
    print (sum (map ((\(r,g,b) -> r*g*b) . getMinRequired) (parseGames games)))

